site/_guides/gatekeeper.md

```diff
@@ -0,0 +1,406 @@
+gatekeeper-controller-manager-7cbc758844-xv9jn   1/1     Running   0          40m
+```
+
+If installation was successful, all pods should reach `Running` status shortly.
+
+## Configure Gatekeeper
+
+### Background
+
+Gatekeeper uses the [OPA Constraint Framework](https://github.com/open-policy-agent/frameworks/tree/master/constraint) to define and enforce policies.
+This framework has two key types: `ConstraintTemplate` and `Constraint`.
+A `ConstraintTemplate` defines a reusable OPA policy, along with the parameters that can be passed to it when it is instantiated.
+When a `ConstraintTemplate` is created, Gatekeeper automatically creates a custom resource definition (CRD) to represent it in the cluster.
+
+A `Constraint` is an instantiation of a `ConstraintTemplate`, which tells Gatekeeper to apply it to specific Kubernetes resource types (e.g. `HTTPProxy`) and provides any relevant parameter values.
+A `Constraint` is defined as an instance of the CRD representing the associated `ConstraintTemplate`.
+
+We'll now look at some examples to make these concepts concrete.
+
+### Configure resource caching
+
+First, Gatekeeper needs to be configured to store all `HTTPProxy` resources in its internal cache, so that existing `HTTPProxy` resources can be referenced within constraint template policies.
+This is essential for being able to define constraints that look across all `HTTPProxies` -- for example, to verify FQDN uniqueness.
+
+Create a file called `config.yml` containing the following YAML:
+
+```yaml
+apiVersion: config.gatekeeper.sh/v1alpha1
+kind: Config
+metadata:
+  name: config
+  namespace: "gatekeeper-system"
+spec:
+  sync:
+    syncOnly:
+      - group: "projectcontour.io"
+        version: "v1"
+        kind: "HTTPProxy"
+```
+
+Apply it to the cluster:
+
+```bash
+$ kubectl apply -f config.yml
+```
+
+### Configure HTTPProxy validations
+
+The first constraint template and constraint that we'll define are what we'll refer to as a **validation**.
+These are rules for `HTTPProxy` specs that Contour universally requires to be true.
+In this example, we'll define a constraint template and constraint to enforce that all `HTTPProxies` must have a unique FQDN.
+
+Create a file called `unique-fqdn-template.yml` containing the following YAML:
+
+```yaml
+apiVersion: templates.gatekeeper.sh/v1beta1
+kind: ConstraintTemplate
+metadata:
+  name: httpproxyuniquefqdn
+spec:
+  crd:
+    spec:
+      names:
+        kind: HTTPProxyUniqueFQDN
+        listKind: HTTPProxyUniqueFQDNList
+        plural: HTTPProxyUniqueFQDNs
+        singular: HTTPProxyUniqueFQDN
+  targets:
+    - target: admission.k8s.gatekeeper.sh
+      rego: |
+        package httpproxy.uniquefqdn
+
+        violation[{"msg": msg, "other": sprintf("%v/%v", [other.metadata.namespace, other.metadata.name])}] {
+          got := input.review.object.spec.virtualhost.fqdn
+          other := data.inventory.namespace[_]["projectcontour.io/v1"]["HTTPProxy"][_]
```

Comment:
Since Contour is especially good in multi-tenant situations, I'd be interested to see more mentions of namespaces in the Gatekeeper/Contour presentation.

* are constraints namespace-specific by default; can they be made cluster wide
* how constraints at the cluster level interact with namespace-level ones
* what is the order of precedence
* etc.

Reply:
That's great feedback. AFAIK, both `ConstraintTemplates` and `Constraints` must be cluster-scoped, and there's not any other way in Gatekeeper to specify that constraints only apply to specific namespaces. I'll double-check on that to make sure I'm correct but I believe that's the case.

However, you *can* add logic to the rego to effectively make certain constraints namespace-specific -- you'd add a `namespace` parameter to the `ConstraintTemplate`, and then in the rego, you'd add logic to only check `HTTPProxy` objects in that namespace (if specified).

I'll look at getting an example or two of this set up and covering it in the guide. Thanks again for the review!